src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -273,102 +224,443 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
-            # in this function). This means the variable can not always be
-            # safely parallelised. Example 1:
-            # do j=1, n
-            #    a(1) = b(j)+1
-            #    c(j) = a(1) * 2
-            # enddo
-            # In this case a(1) should be a thread-private scalar.
-            # Example2:
-            # do j=1, n
-            #    if(some_cond)
-            #       a(1) = b(j)
-            #    endif
-            #  enddo
-            # In this case it is not clear if the loop can be parallelised.
-            # So in any case we add the information for the user to decide.
-            self._add_warning("Variable '{0}' is written to, and "
-                              "does not depend on the loop "
-                              "variable '{1}'."
-                              .format(var_info.var_name,
-                                      loop_variable.name))
-            return False
-
-        # Now we have confirmed that all parallel accesses to the variable
-        # are using the same dimension. If there is a case of stencil
-        # access with write operations (read-only has already been tested)
-        # the loop can not be parallelised. E.g. in one j loop:
-        # b(j) = a(j-1) + a(j+1)
-        # a(j) = c(j)
-
-        sym_maths = SymbolicMaths.get()
+        all_write_accesses = var_info.all_write_accesses
+
+        for write_access in all_write_accesses:
+            # We need to compare each write access with any other access,
+            # including itself (to detect write-write race conditions:
+            # a((i-2)**2) = b(i): i=1 and i=3 would write to a(1))
+            for other_access in var_info:
+                if not self.is_array_access_parallelisable(loop_variables,
+                                                           write_access,
+                                                           other_access):
+                    # There is a dependency. Try to give precise error
+                    # messages:
+                    if write_access is other_access:
+                        # The write access has a dependency on itself, e.g.
+                        # a(3) = ...    or a((i-2)**2) = ...
+                        # Both would result in a write-write conflict
+                        str_access = self._language_writer(write_access.node)
+                        self._add_message(f"The write access to "
+                                          f"'{str_access}' causes a "
+                                          f"write-write race condition.",
+                                          DACode.ERROR_WRITE_WRITE_RACE,
+                                          [str_access])
+                    else:
+                        str_write = self._language_writer(write_access.node)
+                        str_other = self._language_writer(other_access.node)
+                        self._add_message(f"The write access to '{str_write}' "
```

Comment:
Now that I notice, these methods that do _add_message but don't have a starting _clean_messages() maybe ought to be private ( _ prefixed), so that the DependencyTool user don't call them directly because it would accumulate to previous error messages.

 In fact maybe only: 'get_all_messages', 'can_loop_be_parallelised' and the `get_*_parameters` should be the public API?

(* I am unsure about partition - discussed in another comment)